lib/sof/cycle/parser.rb

```diff
@@ -0,0 +1,108 @@
+# frozen_string_literal: true
+
+require "active_support/core_ext/hash/keys"
+require "active_support/core_ext/object/blank"
+require "active_support/core_ext/object/inclusion"
+require "active_support/core_ext/hash/reverse_merge"
+require "active_support/isolated_execution_state"
+
+module SOF
+  # This class is not intended to be referenced directly.
+  # This is an internal implementation of Cycle behavior.
+  class Cycle
+    class Parser
+      extend Forwardable
+      PARTS_REGEX = /
+        ^(?<vol>V(?<volume>\d*))? # optional volume
+        (?<set>(?<kind>L|C|W|E) # kind
+        (?<period_count>\d+) # period count
+        (?<period_key>D|W|M|Q|Y)?)? # period_key
+        (?<from>F(?<from_date>\d{4}-\d{2}-\d{2}))?$ # optional from
+      /ix
+
+      def self.dormant_capable_kinds = %w[E W]
+
+      def self.for(notation_or_parser)
+        return notation_or_parser if notation_or_parser.is_a? self
+
+        new(notation_or_parser)
+      end
+
+      def self.load(hash)
```

Comment:
:art: Would it make sense to inject dependencies for more flexibility? For example, here's a slight refactoring:

``` ruby
def self.load(keys: %i[volume kind period_count period_key], defaults: {volume: 1}, **attributes)
  updated_attributes = defaults.merge! attributes.symbolize_keys!
  text = "V#{updated_attributes.values_at(*keys).join}"

  return new text unless updated_attributes[:from_date]

  new [text, "F#{updated_attributes[:from_date]}"].join
end
```

...but even with this refactoring, there are a few issues:

. The `keys` should be a constant.
. The `defaults` should be a constant.
. The `.load` method might not belong on this class? It feels a little out of place, maybe?

Reply:
No, I don't think so. This method is designed to be used in ActiveRecord serialization which depends on a simple dump and load

Reply:
Oh, so for serialization then. That would be a good candidate for serialization objects which would clarify the separation of concerns. Can't tackle that now but would definitely clarify the functionality.